### ISSUE TITLE
LKRG Compatibility / abolish need for userspace loader / /proc info file when initialization succeed

### ISSUE THREAD

**Issue by adrelanos** [0xsirus/tirdad#5]
tirdad and [LKRG](https://www.whonix.org/wiki/Linux_Kernel_Runtime_Guard_LKRG) are currently incompatible. The tirdad module should load before LKRG. This is because LKRG detects modifications made by tirdad as a compromise of kernel integrity.

Loading tirdad using the standard way with `/lib/modules-load.d/` isn't possible because it requires its userspace loader. If the loader would not be required, perhaps `/lib/modules-load.d/` in combination with `/etc/modprobe.d` [settings for Linux module dependencies](https://stackoverflow.com/questions/29717761/how-do-i-define-dependency-among-kernel-modules) could use soft dependencies. LKRG using `MODULE_SOFTDEP("post: module2")` or tirdad using `MODULE_SOFTDEP("pre: p_lkrg")`.

`/etc/modprobe.d` feature to run arbitrary commands (the loader) through `/etc/modprobe.d` syntax `install modulename command` in combination with `/lib/modules-load.d/`  also does not work due to perhaps systemd issue. -> https://github.com/systemd/systemd/issues/14459

I've also attempted to not use `/lib/modules-load.d/` / `/etc/modprobe.d` at all and load both kernel modules, tirdad and LKRG through systemd unit files. First start tirdad. And tirdad systemd unit file saying `Before=lkrg-dkms.service` and LKRG systemd unit file saying `After=tirdad.service`. That doesn't work either. Systemd is "too fast". After running `modprobe` the script finishes and it is considered done by systemd (how could it know any better). But tirdad needs a split second to finish initialization. (`Installing tirdad hook succeeded.`) LKRG runs before that, then seeing the kernel modification and then complaining about that.

There is currently no clean way to check when `Installing tirdad hook succeeded.` is done. Well, the LKRG loader could parse the output of `dmesg` or tirdad loader could do that and until `Installing tirdad hook succeeded.` happened but that is a quite hacky solution. 

Long story short. Possible solutions:

* Could you please consider removing need for any user space loader? Figure out random key and kernel address form inside the kernel module? (preferred)
* And status file in `/proc` (or `/sys`?) creating when `Installing tirdad hook succeeded.` happened?

**Comment by 0xsirus** [comment]
Ok. I don't know when I can get back to this. But I'll consider working on either of the options you mentioned.

**Comment by madaidan** [comment]
What's the point of the userspace loader anyway? To get a kernel symbol?

Kernel modules should be able to get that with `kallsyms_lookup_name` without needing `/proc/kallsyms`.

Seems like `kallsyms_lookup_name("secure_tcp_seq")` should do what you want.

**Comment by 0xsirus** [comment]
The reason that I initially used a user space loader is twofold. Firstly user space solutions are more consistent during time. If you have a good solution in user space, you can well expect that it continues to work reliably as it should in future. We don't have access to kernel memory in user space and this is the part that I had to go into the kernel space to deal with. The rest of the job could be done in user space.
Second reason is that at the time, I didn't know the presence of the user space loader might be problematic in the Whonix setup. But given other protection measures in place, it seems that I had to eventually get rid of it.
After reading your message I dug a little bit in the kernel source and changed the module and removed the loader. I hope that given the new changes it will still satisfy my first reason that I mentioned above. I say hope and not believe because I haven't read the implementation of those new kernel functions that I use to make sure they will consistently work in different situations. So as you see the maintenance of the project gets harder the more your code fares in kernel space.
However with all that said, after testing tirdad in both 4 and 5 kernels and skimming through the aforementioned new functions, I can confirm that currently it works without any problem and the result stays correct and reasonable. I tested the new code exactly the same way I had already done and published in my blog. All good so far and the package is ready be upgraded.
Thank you for your contribution.

**Comment by adrelanos** [comment]
Even without the userspace loader, I think tirdad may still not be LKRG compatible? I will try once compilation issues are sorted out. [1] Does `modprobe` block until `Installing tirdad hook succeeded.` happened?

* If yes: LKRG compatible
* If not: probably not LKRG compatible

I will experiment with `/etc/modprobe.d/30_lkrg.conf`:

```
## Load tirdad before LKRG so LKRG does not judge tirdad to be malicious.
softdep p_lkrg pre: tirdad
```

But I guess that would require tirdad to do the proper signaling to the kernel somehow that initialization completed?

----

[1]

```
user@host:~/tirdad$ make
cd module; make -C /lib/modules/4.19.0-6-amd64/build M=/home/user/tirdad/module
make[1]: Entering directory '/usr/src/linux-headers-4.19.0-6-amd64'
  CC [M]  /home/user/tirdad/module/tirdad.o
In file included from /usr/src/linux-headers-4.19.0-6-common/include/linux/bitmap.h:9,
                 from /usr/src/linux-headers-4.19.0-6-common/include/linux/cpumask.h:12,
                 from /usr/src/linux-headers-4.19.0-6-common/arch/x86/include/asm/cpumask.h:5,
                 from /usr/src/linux-headers-4.19.0-6-common/arch/x86/include/asm/msr.h:11,
                 from /usr/src/linux-headers-4.19.0-6-common/arch/x86/include/asm/processor.h:21,
                 from /usr/src/linux-headers-4.19.0-6-common/arch/x86/include/asm/cpufeature.h:5,
                 from /usr/src/linux-headers-4.19.0-6-common/arch/x86/include/asm/thread_info.h:53,
                 from /usr/src/linux-headers-4.19.0-6-common/include/linux/thread_info.h:38,
                 from /usr/src/linux-headers-4.19.0-6-common/arch/x86/include/asm/preempt.h:7,
                 from /usr/src/linux-headers-4.19.0-6-common/include/linux/preempt.h:81,
                 from /usr/src/linux-headers-4.19.0-6-common/include/linux/spinlock.h:51,
                 from /usr/src/linux-headers-4.19.0-6-common/include/linux/seqlock.h:36,
                 from /usr/src/linux-headers-4.19.0-6-common/include/linux/time.h:6,
                 from /usr/src/linux-headers-4.19.0-6-common/include/linux/stat.h:19,
                 from /usr/src/linux-headers-4.19.0-6-common/include/linux/module.h:10,
                 from /home/user/tirdad/module/tirdad.c:8:
In function ‘memset’,
    inlined from ‘hook_init’ at /home/user/tirdad/module/tirdad.c:176:2:
/usr/src/linux-headers-4.19.0-6-common/include/linux/string.h:330:3: error: call to ‘__write_overflow’ declared with attribute error: detected write beyond size of object passed as 1st parameter
   __write_overflow();
   ^~~~~~~~~~~~~~~~~~
make[4]: *** [/usr/src/linux-headers-4.19.0-6-common/scripts/Makefile.build:315: /home/user/tirdad/module/tirdad.o] Error 1
make[3]: *** [/usr/src/linux-headers-4.19.0-6-common/Makefile:1534: _module_/home/user/tirdad/module] Error 2
make[2]: *** [Makefile:146: sub-make] Error 2
make[1]: *** [Makefile:8: all] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-4.19.0-6-amd64'
make: *** [Makefile:5: all] Error 2
```

**Comment by 0xsirus** [comment]
Fixed the compile error.

> Does modprobe block until `Installing tirdad hook succeeded.` happened?

The installer, whether it be `modprobe` or `insmod`, does wait for the initialize function to complete and reads its return value. `hook_init` has been registered as the initializer function so the installer correctly waits for it to complete before exiting.
For example if you have `module1` with `f1()` as the init function and `module2` with `f2()` as the init function and then you run the following commands sequentially:
#`modprobe module1`
#`modprobe module2`

this is what happens: `module1` is loaded into memory and `f1()` is called. `modprobe` waits for `f1()` to return. `modprobe` reads the return value and interprets it to check whether an error has occured or not. `modprobe` exits and next line is executed. The same thing will happen for `module2` as well.

**Comment by adrelanos** [comment]
This works great now! Thanks a lot!
